wgpu-hal/src/dx12/mod.rs

```diff
@@ -147,104 +175,102 @@ impl D3D12Lib {
+    pub fn debug_interface1(&self) -> Result<Dxgi::IDXGIInfoQueue, crate::DeviceError> {
         // Calls windows::Win32::Graphics::Dxgi::DXGIGetDebugInterface1 on dxgi.dll
         type Fun = extern "system" fn(
             flags: u32,
             riid: *const windows_core::GUID,
             pdebug: *mut *mut core::ffi::c_void,
         ) -> windows_core::HRESULT;
-        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"DXGIGetDebugInterface1") }?;
+        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"DXGIGetDebugInterface1\0") }?;
 
-        let mut result__ = core::ptr::null_mut();
-        Ok((func)(0, &Dxgi::IDXGIInfoQueue::IID, &mut result__)
-            .and_then(|| unsafe { windows_core::Type::from_abi(result__) }))
-    }
+        let mut result__ = None;
 
-    pub fn create_factory1(
-        &self,
-    ) -> Result<windows::core::Result<Dxgi::IDXGIFactory1>, libloading::Error> {
-        // Calls windows::Win32::Graphics::Dxgi::CreateDXGIFactory1 on dxgi.dll
-        type Fun = extern "system" fn(
-            riid: *const windows_core::GUID,
-            ppfactory: *mut *mut core::ffi::c_void,
-        ) -> windows_core::HRESULT;
-        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"CreateDXGIFactory1") }?;
+        (func)(0, &Dxgi::IDXGIInfoQueue::IID, <*mut _>::cast(&mut result__))
+            .ok()
+            .into_device_result("debug_interface1")?;
 
-        let mut result__ = core::ptr::null_mut();
-        Ok((func)(&Dxgi::IDXGIFactory1::IID, &mut result__)
-            .and_then(|| unsafe { windows_core::Type::from_abi(result__) }))
+        result__.ok_or(crate::DeviceError::Unexpected)
     }
 
-    pub fn create_factory2(
+    /// Will error with crate::DeviceError::Unexpected if DXGI 1.4 is not available.
+    pub fn create_factory4(
         &self,
         factory_flags: Dxgi::DXGI_CREATE_FACTORY_FLAGS,
-    ) -> Result<windows::core::Result<Dxgi::IDXGIFactory4>, libloading::Error> {
+    ) -> Result<Dxgi::IDXGIFactory4, crate::DeviceError> {
         // Calls windows::Win32::Graphics::Dxgi::CreateDXGIFactory2 on dxgi.dll
         type Fun = extern "system" fn(
             flags: Dxgi::DXGI_CREATE_FACTORY_FLAGS,
             riid: *const windows_core::GUID,
             ppfactory: *mut *mut core::ffi::c_void,
         ) -> windows_core::HRESULT;
-        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"CreateDXGIFactory2") }?;
+        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"CreateDXGIFactory2\0") }?;
+
+        let mut result__ = None;
 
-        let mut result__ = core::ptr::null_mut();
-        Ok(
-            (func)(factory_flags, &Dxgi::IDXGIFactory4::IID, &mut result__)
-                .and_then(|| unsafe { windows_core::Type::from_abi(result__) }),
+        (func)(
+            factory_flags,
+            &Dxgi::IDXGIFactory4::IID,
+            <*mut _>::cast(&mut result__),
         )
+        .ok()
```

Comment:
I should report over at `windows-rs` that the naming choice of this function is less than ideal. I keep mistaking it for https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.ok which _discards_ `Error` information by turning `Result` into `Option` which instantly triggers during review.